components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -246,6 +254,12 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
 		if err != nil {
 			return admission.Errored(http.StatusInternalServerError, err)
 		}
+
+		// Check Imagestream Info
+		err = SetContainerImageFromRegistry(ctx, w.Client, notebook, log)
```

Comment:
thinking about the future, when rhoai is going to be using upstream kubeflow notebooks 2.0, how is this going to work? will we require patched kubeflow notebooks with this?

Reply:
This part of code is on odh-nbc side, the upgrade to 2.0 will not affect it.

Reply:
But odh notebook controller is a fork of the kubeflow notebook controller, no? So upgrading to 2.0 means we'd have to rebase all custom changes on top of the 2.0 controller. The more changes we make, the harder this is to maintain.

>  I worry that letting dashboard pass in nonsense (:2024.1) and fixing it in notebooks controller is a mistaken architectural decision.

I still worry about this. The most reasonable way to get a fix to 2.8.z release is imo to fix the `:2024.1` thing in dashboard, not this.

Reply:
@jiridanek this is what I was thinking, too, with regards to how to handle the image-field resolution. I think this folder odh-notebook-controller is not part of upstream community kubeflow. Still, I'd advocate for handling this via image content change trigger annotation (set via dashboard during assemble and update) and a new dashboard GUI slider per notebook like "make long-running" or "do not restart of image changes" that sets paused: True

But are there plans to instead construct the image change trigger annotation in odh notebook controller, possibly with paused false or true, depending on a to be defined notebook annotation like "long-running" or "apply image changes immediately" instead of filling the container image field value, in future? https://docs.openshift.com/container-platform/4.12/openshift_images/triggering-updates-on-imagestream-changes.html
My reasoning is: do as little as necessary programatically yourself and use as many features that the platform provides as possible. With the ose-oouth-proxy sidecar and istio, there was no other way than to manually handle all possibilities, but with the trigger annotation, a lot would be handled by it.

Reply:
just put imagestreamname:imagestreamtag as the value for env var JUPYTER_IMAGE in odh-dashboard directly.
dashboard code needs to change so no dependency on dockerImageRepository field of ImageStream (empty when no internal registry present) in imageUrl variable, thereby no dependency on internal Docker registry when spawning images that are coming from an imageStream.
See my latest comment https://github.com/opendatahub-io/kubeflow/pull/329#issuecomment-2128673252

Reply:
@harshad16 @atheo89 I still think it'd be best to go with the image change trigger annotation, letting openshift itself update the image-field accordingly, possibly in odh dashboard having a GUI field (checkmark, slider) per workbench named "no image updates, long-running" that sets image change trigger value paused: False (default) to paused: True, ensuring the workbench/pod/container does not restart when a floating docker.from image tag leads to changing digests.